main.go

```diff
@@ -179,8 +192,17 @@ func (d *Downloader) DownloadWithContext(ctx context.Context, urls ...string) <-
 				s.Error = err
 				return
 			}
+			ds, err := d.getContentSizeHeader(ctx, u)
```

Comment:
I believe we could split this into two parts:

1. Getting and filling the file size through HEAD request + emitting this first DownloadStatus (`DownloadedFileBytes = 0`)
2. Download and fill `DownloadedFileBytes`

How does that sound, @cuducos and @vmesel ?

Reply:
Sorry for the dumb question @danielfireman, how would you split it? Setting the value on the first part and then retrieving it when updating the channel?

Reply:
Not dumb at all, @vmesel. Please find my suggestion

```go
func emitContentSize(ctx Context, url, path string, ch chan DownloadStatus) error {                       
                        s := DownloadStatus{
                                  URL: url,
                                  DownloadedFilePath = path,
                                   DowloadedFileSize = 0,
                         }
                        defer func() { ch <- s }()
                        ds, err := d.getContentSizeHeader(ctx, url)
                        if err != nil {
				s.Error = err
				return 
	                }
                        s.FileSizeBytes = ds
}

for _, u := range urls {
		wg.Add(1)
		go func(u string) {
                        defer wg.Done()
                        path = filepath.Join(os.TempDir(), filepath.Base(u))
                        if err := emitContentSize(ctx, u, path, ch); err != nil {
                                  return
                        }

			s := DownloadStatus{URL: u, DownloadedFilePath = path}
                         defer func() { ch <- s }()
			b, err := d.downloadFile(ctx, u)
			if err != nil {
				s.Error = err
				return
			}
                         ...
```

Reply:
Just checking: I think this example is creating two `DownloadStatus` and the second one will miss the total file size. I guess. 

If that's the case, maybe changing the signature from `emitContentSize(ctx Context, url, path string, ch chan DownloadStatus)` to `emitContentSize(ctx Context, s *DownloadStatus, ch chan DownloadStatus)` might help:
1. create `s := DownloadStatus{URL: u, DownloadedFilePath = path}`
2. send it to `emitContentSize` which gets the size, uses the pointer to save the size, and sends the message through the channel

Reply:
>Just checking: I think this example is creating two DownloadStatus and the second one will miss the total file size. I guess.

Good point!

> If that's the case, maybe changing the signature from emitContentSize(ctx Context, url, path string, ch chan DownloadStatus) to emitContentSize(ctx Context, s *DownloadStatus, ch chan DownloadStatus) might help:

I would suggest simply returning the total download size. What do you think?

Reply:
That'd work too and is simpler : )